### PR TITLE
Update _inferred_steps when recreating iterator.

### LINKED ISSUE #41019: 'fit' takes a dataset with increasing batch size
**System information**
- TensorFlow version (you are using): 2.2
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
I have a custom model to train a text classifier. So far (tensorflow 2.1) we overwrote all important methods of `tf.keras.models.Model`, e.g. `fit`, `predict` etc. One of the reasons for that was that we wanted to adapt the batch size of our dataset every epoch. So, during the first epoch we would have, for example, a batch size of 16, in the next iteration the batch size would increase to 20, and so on. Until we reach an upper limit.

Here is the method that generates our tf.Dataset. We would call this method every epoch to get a dataset with the desired batch size.
```
def as_tf_dataset(
    self, batch_size: int, batch_strategy: Text = SEQUENCE, shuffle: bool = False
) -> tf.data.Dataset:
    """Create tf dataset."""

    shapes, types = self._get_shapes_types()

    return tf.data.Dataset.from_generator(
        lambda batch_size_: self._gen_batch(batch_size_, batch_strategy, shuffle),
        output_types=types,
        output_shapes=shapes,
        args=([batch_size]),
    )
```

With Tensorflow 2.2. it is possible to just overwrite `train_step` and `test_step` instead of `fit` etc (we followed this [guide](https://keras.io/guides/customizing_what_happens_in_fit/)). This would simplify our code quite a bit and would allow us to use callbacks. 
However, when we would overwriting `train_step` instead of `fit` we would loose the possibility to control the dataset, e.g. adapt the batch size of the dataset every epoch. `fit` just takes a dataset and with the help of a `DataHandler` it create a new iterator over that dataset every epoch. 

I am not sure how to solve this, but maybe we could allow users to define a custom `DataHandler` that takes care of the dataset generation. 

**Who will benefit with this feature?**
Any users that use a `tf.keras.models.Model`.

### PR DESCRIPTION
**Pull request by tabergma**
closes https://github.com/tensorflow/tensorflow/issues/41019

In case your data generator returns different number of batches per epoch, we need to update the `inferred_steps` after recreating the iterator over the dataset. This allows users to create data generators that modify the datasets and update the batch size every epoch, for example, using an increasing batch size.

### CODE REVIEW

**Comment by tabergma** [comment]
I have some trouble setting up the tests locally to see the actual failures. Do you have any guide for that? Or can you point me to the actual errors why tests are failing?

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. Please add a unit test to check this change.

**Comment by gbaned** [comment]
@tabergma Can you please check @fchollet's comments and keep us posted ? Thanks!

**Comment by tabergma** [comment]
@gbaned @fchollet Thanks for the review! Not sure when I have time to add the unit test as I have a newborn. I'll keep you posted.

**Comment by gbaned** [comment]
@tabergma  Any update on this PR? Please. Thanks!

**Comment by tabergma** [comment]
I did not find time to add a test yet, try to do it in the next week.

**Comment by tabergma** [comment]
I started looking into it, but had some problems running the tests locally. Not sure when I'll have time to finish it.

**Comment by gbaned** [comment]
@tabergma Any update on this PR? Please. Thanks!

**Comment by tabergma** [comment]
Unfortunately, no update. I guess I'm not able to find time any time soon. Will let you know when I start working on it again. Sorry for the delay.

**Comment by tensorflowbutler** [comment]
It has been 19 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by markus-hinsche** [comment]
I would like to follow up with this PR. I work in @tabergma 's team. 
@gbaned @fchollet Would it be ok to re-open this PR? Would you still accept this contribution?

**Comment by gbaned** [comment]
Hi @markus-hinsche  It looks like this PR relates to the Keras component. Please submit it to the github.com/keras-team/keras repository instead. Thankyou.
@fchollet, @qlzh727
